src/plugins/lua/replies.lua

```diff
@@ -94,18 +98,42 @@ local function replies_check(task)
     return nil
   end
 
-  local function add_to_global_replies_set(params, sender_key)
+  local function add_to_global_replies_set(params, sender_key, global_key)
     lua_util.debugm(N, task, 'Adding recipients %s to global replies set', params)
 
+    local function zrem_cb(err, data)
+      if err ~= nil then
+        rspamd_logger.errx(task, 'failed to remove extra senders from global replies set with error: %s', err)
+        return
+      end
+      rspamd_logger.infox(task, 'if necessary, additional senders were deleted from global replies set')
+    end
+
     local function zadd_global_set_cb(err, data)
       if err ~= nil then
         rspamd_logger.errx(task, 'failed to add sender %s to global replies set with error: %s', sender_key, err)
         return
       end
       rspamd_logger.infox(task, 'added sender %s to global set with code: %s', sender_key, data)
+
+      local redis_script_zadd_global = [[
+      local global_size = redis.call('ZCARD', KEYS[1])
+      if global_size > {= mgs =} then
+        redis.call('ZREMRANGEBYRANK', KEYS[1], '{= mgs =}', global_size)
+      end
+      ]]
+
+      local set_script_zadd_global = lua_util.jinja_template(redis_script_zadd_global,
+              { mgs = settings['max_global_size'] })
+      local zadd_id = lua_redis.add_redis_script(set_script_zadd_global, redis_params)
```

Comment:
This code part must be done in the configuration part of the module, not in the callback. I would also suggest to remove `ZADD` invocation and do all things: `ZCARD` -> `ZREM` -> `ZADD` in one Redis script.